File: src/Day19.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@SuppressWarnings({"ExplicitToImplicitClassMigration", "RedundantSuppression"})
public class Day19 {

  public static class Trie<T> {

    private final TrieNode<T> root = new TrieNode<>();

    public static class TrieNode<T> {
      Map<T, TrieNode<T>> children = new HashMap<>();
      boolean isEndOfWord = false;

      @Override
      public String toString() {
        return "(children: " + children + ", isEnd: " + isEndOfWord + ")";
      }
    }

    public void insert(Iterable<T> seq) {
      TrieNode<T> current = root;
      for (T token : seq) {
        current = current.children.computeIfAbsent(token, _ -> new TrieNode<>());
      }
      current.isEndOfWord = true;
    }

    public boolean search(Iterable<T> seq) {
      TrieNode<T> current = root;
      for (T token : seq) {
        current = current.children.getOrDefault(token, null);
        if (current == null) {
          return false;
        }
      }
      return current.isEndOfWord;
    }

    @Override
    public String toString() {
      return "Trie(" + root + ")";
    }
  }

  private static <T> long dfsCount(Trie<T> trie, List<T> iter, Map<List<T>, Long> cache) {
    if (iter.isEmpty()) {
      cache.putIfAbsent(iter, 1L);
      return 1L;
    }
    if (cache.containsKey(iter)) {
      return cache.get(iter);
    }
    long count = 0L;
    for (int i = 0; i < iter.size(); i++) {
      List<T> prefix = iter.subList(0, i + 1);
      if (trie.search(prefix)) {
        count += dfsCount(trie, iter.subList(i + 1, iter.size()), cache);
      }
    }
    cache.putIfAbsent(iter, count);
    return count;
  }

  private static Pair<Integer, Long> solve(List<String> input) {
    String[] flags = input.stream()
        .filter(s -> s.contains(","))
        .findFirst()
        .orElse("")
        .split(", ");
    Trie<Character> trie = new Trie<>();
    for (String flag : flags) {
      trie.insert(flag.chars().mapToObj(c -> (char) c).toList());
    }
    List<String> reqs = input.stream()
        .skip(1)
        .filter(req -> req.matches(".*[wubrg].*"))
        .toList();
    var count = 0;
    var sum = 0L;
    for (String req : reqs) {
      var tmp = dfsCount(trie, req.chars().mapToObj(c -> (char) c).toList(), new HashMap<>());
      if (tmp > 0) {
        count++;
        sum += tmp;
      }
    }
    return new Pair<>(count, sum);
  }

  public static void main(String[] ignoredArgs) throws IOException {
    List<String> testInput = readInput("19t");
    Pair<Integer, Long> testResult = solve(testInput);
    if (testResult.first != 6) {
      throw new AssertionError("Expected 6 but got " + testResult.first);
    }
    List<String> input = readInput("19");
    long start = System.nanoTime();
    Pair<Integer, Long> result = solve(input);
    long end = System.nanoTime();
    System.out.println(result.first);
    System.out.println(result.second);
    System.out.println("Took " + (end - start) / 1_000_000 + " ms for both parts");
  }

  public static List<String> readInput(String fileName) throws IOException {
    return Files.readAllLines(Paths.get("src", fileName + ".txt"));
  }

  private record Pair<F, S>(F first, S second) {
  }
}

File: src/Day19.kt
import kotlin.time.measureTimedValue

class Trie<T> {

    private val root = TrieNode<T>()

    class TrieNode<T> {
        val children = hashMapOf<T, TrieNode<T>>()
        var isEndOfWord = false
        override fun toString(): String {
            return "(children: $children, isEnd: $isEndOfWord)"
        }
    }

    fun insert(seq: Iterable<T>) {
        var current = root
        for (token in seq) {
            current = current.children.getOrPut(token) { TrieNode() }
        }
        current.isEndOfWord = true
    }

    fun search(seq: Iterable<T>): Boolean {
        var current = root
        for (token in seq) {
            current = current.children[token] ?: return false
        }
        return current.isEndOfWord
    }

    override fun toString(): String = "Trie($root)"
}

fun main() {

    fun <T> Trie<T>.dfsCount(iter: List<T>, cache: HashMap<List<T>, Long> = hashMapOf()): Long {
        if (iter.isEmpty()) return 1L.also { cache.compute(iter) { _, v -> v ?: 1 } }
        if (iter in cache) return cache[iter]!!
        var count = 0L
        for (i in iter.indices) {
            val prefix = iter.subList(0, i + 1)
            if (search(prefix)) {
                count += dfsCount(iter.subList(i + 1, iter.size), cache)
            }
        }
        return count.also { cache.compute(iter) { _, v -> v ?: count } }
    }

    fun solve(input: List<String>): Pair<Int, Long> {
        val flags = input.first { it.contains(',') }.split(", ")
        val trie = Trie<Char>()
        flags.forEach { trie.insert(it.toList()) }
        println(trie.toString())
        val reqs = input.drop(1).filter { it.contains(Regex("[wubrg]")) }
        val res = reqs.map { req -> trie.dfsCount(req.toList()) }
        return res.count { it > 0 } to res.sum()
    }

    val testInput = readInput("19t")
    check(solve(testInput).first == 6) { "Expected 6 but got ${solve(testInput).first}" }
    val input = readInput("19")
    measureTimedValue {
        solve(input)
    }.also {
        println(it.value.first)
        println(it.value.second)
        println("Took ${it.duration} for both parts")
    }
}